### PR TITLE
MDS

### PR DESCRIPTION
**Pull request by mcreixell**
MDS code to find associations between kinase profiles and cluster motifs.

### CODE REVIEW

**Comment by mcreixell** [comment]
@aarmey This is what I'm getting:
[mds.pdf](https://github.com/meyer-lab/resistance-MS/files/5588205/mds.pdf)

However I'm seeing that with this current set up, this plot changes every time I run this code (`ofd.ipynb`, 1st section). Maybe there's something wrong in how I'm using `itertools.permutations`? 

https://github.com/meyer-lab/resistance-MS/blob/e575cacb81c769ecaabe29a5c58bba28f74afe02/msresist/clustering.py#L132-L141

Also, I've seen in examples wherein after using MDS people run [PCA](https://scikit-learn.org/stable/auto_examples/manifold/plot_mds.html#sphx-glr-auto-examples-manifold-plot-mds-py). Is this also the case here or there's no need since MDS already reduces the variance in the distances into 2 components?

**Comment by aarmey** [comment]
Yeah, I'm not sure you're using permutations correctly. Why not just write a loop inside a loop? It'll already be very fast.

MDS is randomly initialized, so it's no surprising that it would vary from run to run. It looks like sklearn is using PCA to rotate the components so that they come out in a consistent orientation between MDS and NMDS. That's fine to do, but shouldn't change the results too much.

**Comment by mcreixell** [comment]
It's true that by using the mds fitting to initialize the nmds fitting:
 `npos = nmds.fit_transform(res, init=pos)`

and then using PCA, the results are reproducible from run to run. What is encouraging is that ABL. although along with several other kinases, has the same PC2 weight than cluster 4, which has the ABL motif. Looks like the profiling of SH2 domains are very unique compared to the rest of PSPL data...  

Screen Shot 2020-11-24 at 11 45 53 AM

File: msresist/clustering.py [hunk 1]
```diff
@@ -129,18 +131,40 @@ def pssms(self, PsP_background=False):
...
+        pos = mds.fit(res).embedding_
+
+        nmds = MDS(n_components=2, metric=False, max_iter=3000, eps=1e-12,
+                            dissimilarity="precomputed", random_state=seed, n_jobs=1,
+                            n_init=1)
+        npos = nmds.fit_transform(res, init=pos)
+
+        clf = PCA(n_components=2)
+        npos = clf.fit_transform(npos)
+        varExp = np.round(clf.explained_variance_ratio_, 2)
```
Comment (aarmey) [inline msresist/clustering.py:159]: The PCA is of the already reduced MDS space, so you can't provide a variance explained.

Also, you might want to look at components 3 and 4. Those might separate the remaining kinases bunched together.

Comment (mcreixell) [reply msresist/clustering.py:159]: Ok, variance explained removed.

This is how it looks like fitting MDS, NMDS, and PCA with 4 components:

Screen Shot 2020-11-24 at 2 08 01 PM

**Comment by mcreixell** [comment]
- In this last commit I'm updating the figure files to show this MDS analysis instead of the heat maps to identify upstream kinases. I still have to play with the kinase/cluster labels to show only those that we care about in each case but I will do this in the next PR. 
- Also, if you remember, during our last UO1 most of the logistic regression cluster coefficients where meaningful when predicting mutations. Switching to full L1 instead of elastic net improved prediction and decreased the number of clusters with large coefficients. Especially in the case of TP53, where we only have one weighted cluster.

**Review by aarmey — APPROVED** [review]
Excellent.
